server/api/region.go

```diff
@@ -313,15 +337,48 @@ func newRegionsHandler(svr *server.Server, rd *render.Render) *regionsHandler {
 	}
 }
 
-func convertToAPIRegions(regions []*core.RegionInfo) *RegionsInfo {
-	regionInfos := make([]RegionInfo, len(regions))
+// regionsToBytes converts regions to bytes, which is `RegionsInfo`'s json format.
+// It is used to reduce the cost of json serialization.
+func regionsToBytes(ctx context.Context, regions []*core.RegionInfo) ([]byte, error) {
+	out := &jwriter.Writer{}
+	out.RawByte('{')
+
+	out.RawString("\"count\":")
+	out.Int(len(regions))
+
+	out.RawString(",\"regions\":")
+	out.RawByte('[')
+	region := &RegionInfo{}
 	for i, r := range regions {
-		InitRegion(r, &regionInfos[i])
-	}
-	return &RegionsInfo{
-		Count:   len(regions),
-		Regions: regionInfos,
+		select {
+		case <-ctx.Done():
+			// Return early, avoid the unnecessary computation.
+			// See more details in https://github.com/tikv/pd/issues/6835
+			return nil, ctx.Err()
+		default:
+		}
+		if i > 0 {
+			out.RawByte(',')
+		}
+		InitRegion(r, region)
```

Comment:
I noticed you keep region for `initRegion`.
If `Buckets` is empty, the last `Buckets` value is retained.

Reply:
good catch